services/app-api/src/resolvers/rate/submitRate.ts

```diff
@@ -0,0 +1,153 @@
+import type { Store } from '../../postgres'
+import type { MutationResolvers } from '../../gen/gqlServer'
+import {
+    setErrorAttributesOnActiveSpan,
+    setResolverDetailsOnActiveSpan,
+} from '../attributeHelper'
+import type { RateFormDataType } from '../../domain-models'
+import { isStateUser } from '../../domain-models'
+import { logError } from '../../logger'
+import { ForbiddenError, UserInputError } from 'apollo-server-lambda'
+import { NotFoundError } from '../../postgres'
+import { GraphQLError } from 'graphql/index'
+import type { LDService } from '../../launchDarkly/launchDarkly'
+
+export function submitRate(
+    store: Store,
+    launchDarkly: LDService
+): MutationResolvers['submitRate'] {
+    return async (_parent, { input }, context) => {
+        const { user, span } = context
+        const { rateID, submitReason, formData } = input
+        const featureFlags = await launchDarkly.allFlags(context)
+
+        setResolverDetailsOnActiveSpan('submitRate', user, span)
+        span?.setAttribute('mcreview.rate_id', rateID)
+
+        // throw error if the feature flag is off
+        if (!featureFlags?.['rate-edit-unlock']) {
+            const errMessage = `Not authorized to submit rate, the feature is disabled`
+            logError('submitRate', errMessage)
+            throw new ForbiddenError(errMessage, {
+                message: errMessage,
+            })
+        }
+
+        // This resolver is only callable by State users
+        if (!isStateUser(user)) {
+            const errMessage = 'user not authorized to submit rate'
+            logError('submitRate', errMessage)
+            setErrorAttributesOnActiveSpan(errMessage, span)
+            throw new ForbiddenError(errMessage)
+        }
+
+        // find the rate to submit
+        const unsubmittedRate = await store.findRateWithHistory(rateID)
+
+        if (unsubmittedRate instanceof Error) {
+            if (unsubmittedRate instanceof NotFoundError) {
+                const errMessage = `A rate must exist to be submitted: ${rateID}`
+                logError('submitRate', errMessage)
+                setErrorAttributesOnActiveSpan(errMessage, span)
+                throw new UserInputError(errMessage, {
+                    argumentName: 'rateID',
+                })
+            }
+
+            logError('submitRate', unsubmittedRate.message)
+            setErrorAttributesOnActiveSpan(unsubmittedRate.message, span)
+            throw new GraphQLError(unsubmittedRate.message, {
+                extensions: {
+                    code: 'INTERNAL_SERVER_ERROR',
+                    cause: 'DB_ERROR',
+                },
+            })
+        }
+
+        const draftRateRevision = unsubmittedRate.draftRevision
+
+        if (!draftRateRevision) {
+            throw new Error(
+                'PROGRAMMING ERROR: Status should not be submittable without a draft rate revision'
```

Comment:
No change needed since this is a pattern in other places but I find it odd we are having to check this kind of stuff in the resolver - would like this handled closer to the database. Would be nice if by the time we get to the resolver we trust our `status` field fully.

Reply:
I believe this is a thing with the types.

```ts
const rateSchema = z.object({
    id: z.string().uuid(),
    createdAt: z.date(),
    updatedAt: z.date(),
    status: statusSchema,
    stateCode: z.string(),
    stateNumber: z.number().min(1),
    // If this rate is in a DRAFT or UNLOCKED status, there will be a draftRevision
    draftRevision: rateRevisionWithContractsSchema.optional(),
    // All revisions are submitted and in reverse chronological order
    revisions: z.array(rateRevisionWithContractsSchema),
})
```

We would need to update `draftRateSchema` like this:
```ts
const draftRateSchema = rateSchema.extend({
    status: z.literal('DRAFT'),
    draftRevision: rateRevisionWithContractsSchema,
    revisions: z.array(rateRevisionWithContractsSchema).min(1),
})
```
 Then export a `DraftRateType` to use in `findRateWithHistory` to return the type based on the status.